modules/imgproc/src/hough.cpp

```diff
@@ -1212,15 +1455,14 @@ class HoughCircleEstimateRadiusInvoker : public ParallelLoopBody
                                 ddata[nzCount] = rbuf[p]; nzCount++;
                             }
                         }
+
+                        nnz = 0;
                     }
                 }
-#endif
 
-                // Estimate best radius
-                for(; j < nzSz; ++j)
+                for(int iz = 0; iz < nnz; ++iz)
```

Comment:
This loop is a "tail processing" of SIMD optimized code. It can't start from zero again.

SIMD loop above is broken too. Also SIMD code is totally optional and can be excluded from sources.

Reply:
Correct me if I'm wrong. I had to study the code to figure out how it worked. The previous code was loading data directly from the `nz` vector. It looped in increments of 4 for SIMD processing. This would leave a remainder of 0 to 3 points in `nz` that still needed to be processed. This is conveniently done by falling through to the non-SIMD loop. 

The new code can't do that, because `nz` is a container that is accessed only through a forward iterator. If the container is matrix-based, it contains flags, not points. The first loop fills the `nzx` and `nzy` arrays. When `nnz == 4` the arrays are loaded into vector registers and processed. When the iterator reaches the end, there will be 0 to 3 remaining points in `nzx` and `nzy`. The second loop is there to process them. The loop starts from 0 because it is an index into the `nzx` and `nzy` arrays. 

I also added an `else` to prevent the SIMD code from falling through into the non-SIMD loop. 

There still is a non-SIMD loop for the case where SIMD is disabled either at compile time or run time. To be honest, I have only been testing the SIMD code. The non-SIMD code should work, but it needs to be tested. 

I don't know why you are saying the SIMD loop is broken. The design was changed to work with a forward iterator instead of a random-addressable vector. It still produces the same results. Maybe there is something I did wrong or could do better.

Reply:
The SIMD code is correct (even it has some non-evident logic) but it's arguable whether it gives us additional performance. I think you can freely exclude it from source.